### ISSUE TITLE
Possible performance improvement for very large index (>100M docs): Transparent huge pages?

### ISSUE THREAD

**Issue by ellenhp** [quickwit-inc/tantivy#2316]
## tl;dr:

I executed `echo 'always' | sudo tee /sys/kernel/mm/transparent_hugepage/enabled` and observed queries that used to take **~80ms** go down to **~35ms** for my index of **172M docs**. But that might just be how I'm using tantivy?

## details

Emphasis added because this is long and I expect people to skim :)

I'm building a geocoder[^1] on top of tantivy, and I'm doing some pretty weird stuff with anonymous memory maps and userfaultfd to get it working with an index stored in object storage. The details aren't important, but I ended up experimenting with explicitly requesting huge pages which planted an idea in my brain for later.

I did some profiling of my downstream application after getting my remote directory implementation working, and noticed requests were taking around **80ms** to return queries with a hot cache. That's with around **172M documents**. That's slower than I'd like, so I looked into why and found this:

![image](https://github.com/quickwit-oss/search-benchmark-game/assets/1936763/91f7ee01-973c-4a3b-b62d-cc2c1fed2a12)

**Image description for accessibility**: screenshot of profiling tool showing that `__memset_avx2_unaligned_erms`  accounts for 77% of samples on a given thread, with downstream frames showing that it calls into the kernel which spends a bunch of time handling a page fault.

After spending way too much time investigating my own code I came to the conclusion that the page faults perf was recording weren't against the index, and I think they're within tantivy. I'm still not sure where [^2] these page faults were occurring, but cave-woman brain decided that huge tables means fewer page faults so less wasted time, so I decided to force transparent huge pages on my system temporarily and saw an immediate **~2.3x** improvement in query latency against my remote index when the cache was hot (no misses, whole thing in physical memory).

I'm curious if this is due to a quirk of the way I'm using tantivy, or if this setting could benefit other users. If it were like a 5% performance improvement I wouldn't bother opening an issue, but I'm observing a full 2x I was wondering if someone could try this and tell me if I'm crazy or not. I tried it with a local index of **1M docs** and it didn't make a difference (cursory glance shows it's 6% slower actually, not sure what the confidence on that number is though). I don't really want to download the full planet index from S3 to test it locally and rule out the cause being some quirk of my userfaultfd directory implementation. I could do that if people can't reproduce this, but I'd prefer to avoid egress fees otherwise.

[^1]: https://github.com/ellenhp/airmail
[^2]: I think it might have something to do with the initialization of the docsets during a query? 200M documents means 200M bits in the bitset, naively, that seems like it could lead to an expensive `__memset_avx2_unaligned_erms` operation? This is purely a guess.

**Comment by adamreichold** [comment]
I would say transparent huge pages can generally be hit or miss and usually become a tuning problem, i.e. I would say that your experience of it helping there but making things worse here is pretty typical. (For example, I could image that TLB pressure is worse in a multi-tenant Cloud environment than locally on a dedicated machine.)

So even though the kernel implementation is continuously improving, the feature never was fully "transparent" and the kernel's heuristics for deciding which base pages to consolidate can help as well as hinder (both due the overhead of deciding in the first place, but also by increasing memory consumption which is then missing elsewhere, e.g. for other kernel caches).

Also note that AFAIK, huge pages still only apply to anonymous memory, i.e. all the stuff handled via file-backed memory maps using `MmapDirectory` will use base pages as the kernel's page cache will manage only those.

> I think it might have something to do with the initialization of the docsets during a query? 200M documents means 200M bits in the bitset, naively, that seems like it could lead to an expensive __memset_avx2_unaligned_erms operation? This is purely a guess.

If I understand things correctly, the bitmaps are materialized only per segment so there should never be a 24 MB bitmap covering the whole 200M-documents-large index or put differently `DocId` is always a within-segment identifier.

**Comment by adamreichold** [comment]
So maybe to add another data point, I tried forcing THP on a set of search benchmarks for our small (< 1 GB) classic (i.e. `MmapDirectory` based) index and I see wins and losses in the single-digit percentages depending on the query.

**Comment by ellenhp** [comment]
My index is on the larger side, and what's unusual about it is that I (foolishly) merged the segments, which could be part of why I'm seeing weird behavior. When I get around to indexing the planet again I'm going to let tantivy handle the segment merging according to its policies. We'll see though, I might end up with huge segments like this again because next time I index I'm going to include OpenAddresses so it'll weigh in around 800M docs. This issue was mostly a question and I think it's answered so we should be able to close it. But I might update this with more information next time I index, if that's ok? I index locally so I can also test MmapDirectory and see if I can reproduce this observation.

**Comment by fulmicoton** [comment]
@ellenhp 
What is the setting for this result? Is it a running server, or is it a CLI, where you map stuff in memory run the query and exit the program?

**Comment by ellenhp** [comment]
It's a server. The results are very repeatable. Every time I hit the search endpoint with a (specific) query that doesn't result I'm an S3 request it's the same thing. Either 82ms+/-5 or 35ms+/-5. To get the profiling results I hit the endpoint 50ish times. 

Edit: single tenant machine though. A desktop.

**Comment by ellenhp** [comment]
I enabled the quickwit feature to get an sstable termdict and my hot-cache queries are down to 2ms, so what I observed earlier must be related to the fst termdict.

**Comment by fulmicoton** [comment]
Just to be sure I understand. You don't mmap the files at each query right?

**Comment by ellenhp** [comment]
Correct. To be more specific though, there are no mmaped files, I do an HTTP HEAD request on server startup with each of files in my remote directory to decide how many pages to request for anonymous memory maps for each, then register a userfaultfd handler for all of the maps so whenever any of those pages fault, I can perform a HTTP range request into S3 or wherever else then copy the correct data into the page. Once the page is mapped, the data sits there until the kernel reclaims it then I'll have to hit S3 again the next time it faults. (at least, if i understand the madvise docs right that's what'll happen)

It's a bit more complicated than that because I request a much bigger chunk than just the page size, but that's the gist of it. I was doing this with safe rust before and it was way too many copies, but the mmap seems to be working out well.

**Comment by fulmicoton** [comment]
Got it. That might be too specific to draw any conclusion about whether this is useful for more normal workload.

Also, the "quickwit" feature should really help with your set up.

**Comment by ellenhp** [comment]
Yeah, with the way the internet works this'll probably come up for someone in a search in a few years and they'll be able to figure out exactly what was happening and if it'll help them. Tentatively think it's to do with the FST termdict though. But who knows. Thanks for the tip, and yeah the quickwit feature is doing wonders.

**Comment by adamreichold** [comment]
> Tentatively think it's to do with the FST termdict though

But isn't that mapped from file-backed storage and hence should not benefit from THP at all?

You say that your custom fetching from S3 uses anonymous memory so I would rather that this is what benefits from THP. Usage of `userfaultfd` might also fit because it will mean larger but fewer page faults to handle over the kernel/userland barrier?

**Comment by ellenhp** [comment]
I had the same results with and without explicitly requesting huge pages. It's possible that THP does something I don't fully understand, but my intuition is that if I'm explicitly requesting huge pages, it should have no effect. Regardless, right now I'm using 4k pages and with/without THP enabled, searches complete in about 2ms assuming no page faults. sstables are magic. Tantivy is a seriously impressive piece of software.

I'm going to leave this THP mystery for the next person, since I have my solution, I think. :)
